### ISSUE TITLE
Support for QR code in explorer

### ISSUE THREAD

**Issue by gregscullard** [hashgraph/hedera-mirror-node-explorer#594]
### Problem

Inputting account numbers in wallets manually can lead to errors, suggestion is to add a QR code for accounts (maybe also contracts) such that they can be read by a wallet application.

Proposal is to add a link to accounts and contracts such that AccountId, EVM address, Alias can be displayed as a QR code.

For IDs (such as accountId and contractId), the QR code being displayed should include an option to include the checksum or not (default with checksum)

### Solution

Add links to IDs and addresses (EVM, Alias) in the explorer that pop up a QR code, if the QR code is for an ID, include an option to enable/disable the checksum in the QR code (default on).

**Comment by lbaird** [comment]
That's true. A QR code already has its own checksum. So it would be fine for the QR code to just encode 0.0.1234 rather than 0.0.2234-abcde. The extra checksum would only be used when a user is manually typing the account number.

**Comment by lbaird** [comment]
Most wallets (such as Hashpack) already show QR codes for account IDs.

**Comment by Liebenfiels** [comment]
**I think QR code in text readable format would be tremendous help**, it work in early days when users have myhbarwallet phone version with QR. QR codes in single text readable standard are broadly used in many networks and Hedera accounts are not longer 4-5 digits in length 
 it will help as it can work as such:
**Explorer via desktop to phone wallet interactions**- user types account ID on desktop explorer where it has all the additonal details of account (balance, last transaction, node staked, check sum) so very unlikely it will not realize the account is wrong as all the other details will not match. When using desktop it is easier to type correctly on physical keyboard than touch screen on mobile phone.
Then I would suggest to include **QR code for accounts with check sum on explorers and wallet** (must be text readable format to be able to confirm with phone camera, not locked wallet specific QR that user cannot read), and account copy button as on blockchair.com explorers for many network.  USer need to understand check sum, hence I would include it.
Wallet would do the check if account ID matches check sum in QR code (account ID + check sum as text readable). The 5 letter check sum adds that extra needed length to add variability for QR, good especially for shorter accounts, it is less likely it will fail and if fails it will fail safe as e.g. account ID may not match check sum.
Then user would grab the QR code with phone without need to re-type in phone wallet. 

Also user can send snapshot of QR code and with account details e.g. 1234 in email/priv message and easier for receiver to scan QR and also check if it matches 1234 that user stated what the QR code is for. 
It will be secondary check for user: scan QR and check if displays 1234 as in email.

**Phone wallet to phone wallet interaction**- wallet to display text readable QR code for account with check sum. Other user scans wallet QR e.g. meet up/conference for direct interactions as with other network. Can you scan now blade wallet QR code with hashscan  phone wallet?
 All wallets must use single standard as we have multiple wallets now in ecosystem.

**Desktop explorer to chrome wallet extension, or phone wallet and explorer accesses via phone**- here again copy button of account on explorer would help as user would type account ID first on explorer where has all the extra information that wallet will not display, then use copy button to paste to wallet.

How many types people still mistype exchange account, it typed first on explorer they would realize  is wrong as e.g not  many transaction, big balance as expected etc. Of course for frequently used account users adds later to favorite.
Even if does not understand how check sum works will realize that typed wrong account by looking on explorer.
e.g.  1234 instead of 1234 as account balance and history seen on explorer will all be different to what expected.

**Comment by Liebenfiels** [comment]
Also please check what scanner type the Hedera wallet was using at time- I still have working version with QR scanner, but the format ideally need to include the newest checksum.  I spoke to Nik S at time why there is no QR code on his explorer he said at a time because Hedera had no check sum (years back) implemented and it will not be secure to put QR for account ID only.

**Comment by Liebenfiels** [comment]
Can we have an update please: 
**Explorers**- there are still no QR code for account with checksum
**Wallets**- some have some don’t,
 but **no uniform single QR standard exist in ecosystem after 5 years of network existence****
Thank you
Michal

**Comment by bguiz** [comment]
A registry/ name service - for example ENS-like - could be one way to solve for this... but I think not so relevant for this particular issue.

> Besides if I type the wrong account Id (0.0.1233 instead of 0.0.1223), the QR code will show the 0.0.1233 including its checksum and that will mean the error in typing the account id in the explorer won't be noticed by the wallet.

**Comment by bguiz** [comment]
In terms of how this could be designed into the UI, I do like the way that [`iancoleman.io/bip39`](https://iancoleman.io/bip39/) does it, where if you click on a field it toggles the visibility of the QR code in the corner.
![image](https://github.com/hashgraph/hedera-docs/assets/1773785/bd65fe9d-5ea9-48c5-a3bd-f8b864e74bca)

(The same site also toggles upon mouse hover, but I'm not a fan, as it tends to result in a bunch of "flicker")

**Comment by Liebenfiels** [comment]
This explorer has good one also please not it is copy button on right, you need both really https://blockchair.com/bitcoin/address/34xp4vRoCGJym3xR7yCVPFHoCNxv4Twseo

**Comment by tonycamero** [comment]
As a builder actively deploying Hedera-based applications in highly-regulated and community-centric markets (e.g., CraftTrust, Civic Stack, The Culture Wallet), I want to voice strong support for this QR code implementation in the explorer.

Many of our users are not crypto-native — they rely on familiar patterns like QR scanning for payments, messaging, or in-person engagement. Account IDs with checksums in a text-readable QR format would drastically reduce user error and improve trust in face-to-face interactions.

This isn’t just about developer convenience — it’s a fundamental UX bridge for adoption in civic, cannabis, and local commerce environments where Scend is already live.

I encourage reconsidering the priority level or at least advancing a community-led QR standard for accounts + contracts. We’d be happy to align our UX implementations accordingly.

**Comment by kpachhai** [comment]
I think this is an amazing idea for sure.

**Comment by Liebenfiels** [comment]
Just to answer on original Greg Scullard “Besides if I type the wrong account Id (0.0.1233 instead of 0.0.1223), the QR code will show the 0.0.1233 including its checksum and that will mean the error in typing the account id in the explorer won't be noticed by the wallet.”

**My original concept was to add alternative brute force resistant way for searching accounts on network / explorers.**
That could be achieved by hashing account ID and hard coding hash output on explorers, while for QR code to include account ID: hash output. 
User would either use current 0.0.XXX.. account format or unique hash- that even if mistyped it would be impossible to find valid account. User could share hash only which would add some privacy when sending out.

**MD5 hash algorithm as best option**- as 32 length similar to Bitcoin addresses and it is fast and well known. Although it is cryptographically broken, but there won’t be any collision for that use case and one can generate MD5 hash in advance. 

During 2025 I generated 25M when there was in fact below 1M account in network and tested for collision (0). 
**At that point any QR code implementation is good**, if to rely on account ID only then I would include with check sum as text input for QR not to confuse.
Below are selected slides from 2022 of concept I shared at a time with some leaders.

![Image](https://github.com/user-attachments/assets/de7a4029-3760-4459-a9dc-1998d33ae7d2)
![Image](https://github.com/user-attachments/assets/42ccb455-a9b8-4f56-8fac-c1fc8658bc75)
